pkg/lib/oauth/handler/handler_token_test.go

```diff
@@ -221,5 +226,175 @@ func TestTokenHandler(t *testing.T) {
+				r := protocol.TokenRequest{
+					"grant_type":    "client_credentials",
+					"client_id":     clientID,
+					"client_secret": "supersecret",
+					"resource":      resourceURI,
+				}
+				ctx := context.Background()
+				resp := handle(ctx, req, r)
+
+				So(resp.Result().StatusCode, ShouldEqual, 200)
+				var body map[string]interface{}
+				err := json.Unmarshal(resp.Body.Bytes(), &body)
+				So(err, ShouldBeNil)
+				So(body["access_token"], ShouldEqual, accessToken)
+				So(body["token_type"], ShouldEqual, "Bearer")
+				So(body["expires_in"], ShouldEqual, 3600)
+				So(body["scope"], ShouldEqual, "read write")
+			})
+
+			Convey("request for subset of scopes", func() {
+				clientResourceScopeService.EXPECT().GetClientResourceByURI(gomock.Any(), clientID, resourceURI).Return(resource, nil)
+				clientResourceScopeService.EXPECT().GetClientResourceScopes(gomock.Any(), clientID, resourceID).Return(allowedScopes, nil)
+
+				accessToken := "access-token-123"
+				tokenService.EXPECT().IssueClientCredentialsAccessToken(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(
+					func(ctx context.Context, opts handler.ClientCredentialsAccessTokenOptions, resp protocol.TokenResponse) error {
+						resp.AccessToken(accessToken)
+						resp.TokenType("Bearer")
+						resp.ExpiresIn(3600)
+						resp.Scope(strings.Join(opts.Scopes, " "))
+						return nil
+					},
+				)
+
+				req, _ := http.NewRequest("POST", "/token", nil)
+				r := protocol.TokenRequest{
+					"grant_type":    "client_credentials",
+					"client_id":     clientID,
+					"client_secret": "supersecret",
+					"resource":      resourceURI,
+					"scope":         "read",
+				}
+				ctx := context.Background()
+				resp := handle(ctx, req, r)
+
+				So(resp.Result().StatusCode, ShouldEqual, 200)
+				var body map[string]interface{}
+				err := json.Unmarshal(resp.Body.Bytes(), &body)
+				So(err, ShouldBeNil)
+				So(body["access_token"], ShouldEqual, accessToken)
+				So(body["token_type"], ShouldEqual, "Bearer")
+				So(body["expires_in"], ShouldEqual, 3600)
+				So(body["scope"], ShouldEqual, "read")
+			})
+
+			Convey("request for invalid scopes", func() {
+				clientResourceScopeService.EXPECT().GetClientResourceByURI(gomock.Any(), clientID, resourceURI).Return(resource, nil)
+				clientResourceScopeService.EXPECT().GetClientResourceScopes(gomock.Any(), clientID, resourceID).Return(allowedScopes, nil)
+
+				req, _ := http.NewRequest("POST", "/token", nil)
+				r := protocol.TokenRequest{
+					"grant_type":    "client_credentials",
+					"client_id":     clientID,
+					"client_secret": "supersecret",
+					"resource":      resourceURI,
+					"scope":         "admin",
+				}
+				ctx := context.Background()
+				resp := handle(ctx, req, r)
+
+				So(resp.Result().StatusCode, ShouldEqual, 400)
+				var body map[string]interface{}
+				err := json.Unmarshal(resp.Body.Bytes(), &body)
+				So(err, ShouldBeNil)
+				So(body["error"], ShouldEqual, "invalid_scope")
```

Comment:
Please also assert the error description.

Reply:
I can add an assertion, but I actually thinks the description is just a text for human to read so it is not expected to be used to compare in any code.

Reply:
Adding assertion on the error description makes sure the code return the expected error. It could catch bugs.